modules/dxp/apps/commerce/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/diagram/AutomappingHandler.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * The contents of this file are subject to the terms of the Liferay Enterprise
+ * Subscription License ("License"). You may not use this file except in
+ * compliance with the License. You can obtain a copy of the License by
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+import {event as d3event, select as d3select, zoom as d3zoom} from 'd3';
+import {fetch} from 'frontend-js-web';
+
+import {ZOOM_VALUES} from './utilities/constants';
+
+class AutomappingHandler {
+	constructor(
+		diagramWrapper,
+		zoomWrapper,
+		imageURL,
+		pinsCSSSelector,
+		updateLabels,
+		updateZoomState
+	) {
+		this._currentScale = 1;
+		this._diagramWrapper = diagramWrapper;
+		this._d3diagramWrapper = d3select(diagramWrapper);
+		this._d3zoomWrapper = d3select(zoomWrapper);
+		this._imageURL = imageURL;
+		this._updateLabels = updateLabels;
+		this._pinBackground = null;
+		this._updateZoomState = updateZoomState;
+		this._zoomWrapper = zoomWrapper;
+		this._handleZoom = this._handleZoom.bind(this);
+		this.rendered = false;
+
+		this._printImage().then(() => {
+			this.rendered = true;
+			this._labels = this._diagramWrapper.querySelectorAll(
+				pinsCSSSelector
+			);
+
+			this._updatePinsState();
+			this._addZoom();
+			this._updateLabels(this._labels);
+		});
+	}
+
+	_addZoom() {
+		this._zoom = d3zoom()
+			.scaleExtent([ZOOM_VALUES[0], ZOOM_VALUES[ZOOM_VALUES.length - 1]])
+			.on('zoom', this._handleZoom);
+
+		this._svg = this._d3diagramWrapper.call(this._zoom);
+	}
+
+	_handleZoom() {
+		this._currentScale = d3event.transform.k;
+
+		if (d3event.sourceEvent) {
+			this._updateZoomState(
+				this._currentScale,
+				d3event.transform,
+				d3event
+			);
+		}
+
+		this._d3zoomWrapper.attr('transform', d3event.transform);
```

Comment:
Good point but unfortunately no. The `d3select` function doesn't return an html element so, every time you want to change some attributes, you need to pass through `d3` which will read in it. 

I could use `this._zoomWrapper` but since I'm dealing with `d3event` I preferred to use d3 for everything

Reply:
![](https://media2.giphy.com/media/XAdbHJywVjF5K/giphy.gif)